src/btree/bt_sync.c

```diff
@@ -62,6 +62,90 @@ __sync_checkpoint_can_skip(WT_SESSION_IMPL *session, WT_PAGE *page)
 }
 
 /*
+ * __sync_dup_walk --
+ *	Duplicate a tree walk point.
+ */
+static inline int
+__sync_dup_walk(WT_SESSION_IMPL *session, WT_REF *walk)
+{
+	bool busy;
+
+	/*
+	 * We already have a hazard pointer, we should generally be able to get
+	 * another one.  That said, we could race with some thread attempting
+	 * to get exclusive access: just continue on to the ordinary write path
+	 * in that case.
+	 */
+#ifdef HAVE_DIAGNOSTIC
+	WT_RET(__wt_hazard_set(session, walk, &busy, __func__, __LINE__));
+#else
+	WT_RET(__wt_hazard_set(session, walk, &busy));
+#endif
+
+	return (busy ? EBUSY : 0);
+}
+
+/*
+ * __sync_evict_page --
+ *	Attempt to evict a page during a checkpoint walk.
+ */
+static int
+__sync_evict_page(
+    WT_SESSION_IMPL *session, WT_REF **walkp, uint32_t flags, bool *skip_walkp)
+{
+	WT_DECL_RET;
+	WT_REF *next, *prev, *to_evict;
+
+	to_evict = *walkp;
+	next = prev = NULL;
+	*skip_walkp = false;
+
+	/*
+	 * Hedge our bets: save a ref on either side of the page we're
+	 * evicting.
+	 *
+	 * !!! Otherwise, the page could split in the meantime, so backing up
+	 * later is not safe.  We allow in-memory splits during checkpoints,
+	 * for example.
+	 */
```

Comment:
I'm glad you put a big comment here, but this seems a bit alarming to me. I don't have another suggestion, and I see that it's not much different to what was happening before - it's just more explicit.